### PR TITLE
forwarder TSO

### PR DESCRIPTION
**Pull request by noaezra**
Support TCP segmentation offload (TSO) in virtio-forwarder. The TCP is a
technique for increasing egress throughput by passing a multipacket
buffer to the NIC, while the NIC splits the buffer into separate
packets. This technique reduces CPU overhead.

### CODE REVIEW

File: lib/netdev-rte-offloads.c [hunk 1]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
+
+    if (l4_proto_id == IPPROTO_TCP) {
+        ol_flags |= (PKT_TX_TCP_SEG | PKT_TX_TCP_CKSUM);
+        m->l2_len = l2_len;
+        m->l3_len = l3_len;
+        m->l4_len = l4_len;
+        m->ol_flags = ol_flags;
+        ret = rte_eth_dev_get_mtu(port_id, &mtu);
+        if (ret < 0) {
+            mtu = 1500;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: 1500 -> ETHERMTU

File: lib/netdev-rte-offloads.c [hunk 2]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
+    struct tcp_hdr *tcp_hdr;
+    uint32_t l2_len = 0;
+    uint32_t l3_len = 0;
+    uint32_t l4_len = 0;
+    uint64_t ol_flags = 0;
+    uint8_t l4_proto_id = 0;
+    uint16_t mtu = 0;
+    int ret = 0;
+
+    if (m->pkt_len < 1400)
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: what is 1400? a finger number? if so, maybe
#define FORWARDER_MIN_TSO_MTU  (ETHERMTU - 100)
and use this const

File: lib/netdev-rte-offloads.c [hunk 3]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
 }
 
+static void
+parse_packet_fields(struct rte_mbuf *m, uint16_t port_id)
+{
+    struct ipv4_hdr *ipv4_hdr;
+    struct ipv6_hdr *ipv6_hdr;
+    struct ether_hdr *eth_hdr;
+    struct tcp_hdr *tcp_hdr;
+    uint32_t l2_len = 0;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: avoid unnecessary initializations, for example if not IPv4/IPv6 packet

File: lib/netdev-rte-offloads.c [hunk 4]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
+    if (m->pkt_len < 1400)
+        return;
+
+    eth_hdr = rte_pktmbuf_mtod(m, struct ether_hdr *);
+    l2_len = sizeof(struct ether_hdr);
+
+    switch (rte_be_to_cpu_16(eth_hdr->ether_type)) {
+    case ETHER_TYPE_IPv4:
+        ipv4_hdr = (struct ipv4_hdr *) ((char *)eth_hdr + l2_len);
+        l3_len = (ipv4_hdr->version_ihl & 0x0f) * 4;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: 4 -> sizeof(uint32_t)

File: lib/netdev-rte-offloads.c [hunk 5]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
+        l4_proto_id = ipv4_hdr->next_proto_id;
+        if (l4_proto_id == IPPROTO_TCP) {
+            tcp_hdr = (struct tcp_hdr *)((char *)ipv4_hdr + l3_len);
+            l4_len = (tcp_hdr->data_off & 0xf0) >> 2;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: >> 2 --> / sizeof(uint32_t)

File: lib/netdev-rte-offloads.c (hunk 4 above)
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: use bit shift (...& 0xf) << 2

File: lib/netdev-rte-offloads.c (hunk 2 above)
Comment (rbaryanai) [reply lib/netdev-rte-offloads.c]: you should always extract the eth and check that (pkt->len - eth_len)  < MTU

File: lib/netdev-rte-offloads.c (hunk 5 above)
Comment (rbaryanai) [reply lib/netdev-rte-offloads.c]: tcp_header->data_off >> 4 gives you the size of the header in 4 bytes units. so 5 means 20 bytes. 
so it should be << 2 . together it is >>2 which is ok (after the & 0xf0)

File: lib/netdev-rte-offloads.c [hunk 6]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
+        ipv6_hdr = (struct ipv6_hdr *) ((char *)eth_hdr + l2_len);
+        l4_proto_id = ipv6_hdr->proto;
+        if (l4_proto_id == IPPROTO_TCP) {
+            l3_len = sizeof(struct ipv6_hdr);
+            tcp_hdr = (struct tcp_hdr *)((char *)ipv6_hdr + l3_len);
+            l4_len = (tcp_hdr->data_off & 0xf0) >> 2;
+            ol_flags |= PKT_TX_IPV6;
+        }
+        break;
+    default:
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: if we have vlan it wont work. which is fine, but it is a problem if don't say anything.

File: lib/netdev-rte-offloads.c [hunk 7]
```diff
@@ -2209,14 +2215,77 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
...
+            tcp_hdr = (struct tcp_hdr *)((char *)ipv4_hdr + l3_len);
+            l4_len = (tcp_hdr->data_off & 0xf0) >> 2;
+            ol_flags |= (PKT_TX_IPV4 | PKT_TX_IP_CKSUM);
+        }
+        break;
+    case ETHER_TYPE_IPv6:
+        ipv6_hdr = (struct ipv6_hdr *) ((char *)eth_hdr + l2_len);
+        l4_proto_id = ipv6_hdr->proto;
+        if (l4_proto_id == IPPROTO_TCP) {
+            l3_len = sizeof(struct ipv6_hdr);
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: this is true only if we don't have extensions. ipv6 extension is a pain, but we need to support in the future.

File: lib/netdev-rte-offloads.c (hunk 5 above)
Comment (elibritstein) [reply lib/netdev-rte-offloads.c]: i think /4 is compiled to >>2, so we can do the code more readable

File: lib/netdev-rte-offloads.c (hunk 4 above)
Comment (elibritstein) [reply lib/netdev-rte-offloads.c]: i think /4 is compiled to >>2, so we can do the code more readable

File: lib/netdev-rte-offloads.c (hunk 7 above)
Comment (elibritstein) [reply lib/netdev-rte-offloads.c]: i agree, but not in the code.
if/when we support VLAN/IPv6-ext, add to the code to support.
currently, it is handled ok (not doing TSO)

File: lib/netdev-rte-offloads.c (hunk 4 above)
Comment (elibritstein) [reply lib/netdev-rte-offloads.c]: and *4 compiled to <<2

Comment (rbaryanai) [reply lib/netdev-rte-offloads.c]: compiler will optimize for sure. I usually don't like to see * and / in a per packet code. not really important.

File: lib/netdev-rte-offloads.c (hunk 7 above)
Comment (rbaryanai) [reply lib/netdev-rte-offloads.c]: If user machine is configured with it, all packets will be dropped and there is not indication why. so it is not handled ok.

**Pull request merged**